spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/ConvertingMessageListener.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2016-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.listener.adapter;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+
+import org.springframework.kafka.listener.AcknowledgingConsumerAwareMessageListener;
+import org.springframework.kafka.listener.AcknowledgingMessageListener;
+import org.springframework.kafka.listener.ConsumerAwareMessageListener;
+import org.springframework.kafka.listener.MessageListener;
+import org.springframework.kafka.support.Acknowledgment;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.converter.MessageConversionException;
+import org.springframework.messaging.converter.MessageConverter;
+import org.springframework.messaging.converter.SimpleMessageConverter;
+import org.springframework.messaging.support.GenericMessage;
+import org.springframework.util.Assert;
+
+/**
+ * A {@link AcknowledgingConsumerAwareMessageListener} adapter that implements
+ * converting received {@link ConsumerRecord} using specified {@link MessageConverter}
+ * and then passes result to specified {@link MessageListener}.
+ *
+ * @param <T> the key type.
+ * @param <U> the value type.
+ * @param <V> the desired value type after conversion.
+ *
+ * @author Adrian Chlebosz
+ * @since 3.0
+ * @see AcknowledgingConsumerAwareMessageListener
+ */
+public class ConvertingMessageListener<T, U, V> implements AcknowledgingConsumerAwareMessageListener<T, U> {
```

Comment:
I look at this from the end-users perspective where they have to think what exactly need to be specified for the `T` and `U` when they really don't care about them since `T` is always going to be an original one and `U` is always going to be converted to `V` independently of `U`.
Therefore I find them redundant and want to avoid questions about them and trying to explain that we need them merely to satisfy an internal code consistency for compiler.
I believe we still can find some hard casting behavior internally even if it could be handled with a `rawtypes` suppression...

Reply:
Changed in eeb9961c